### PR TITLE
fix(ClientRequest): support empty "username" and "password" basic auth values

### PR DESCRIPTION
**Pull request by mikicho**
According to [this test](https://github.com/nock/nock/blob/feaa66fa64d24f95937ef759cdd5a7ca07646f1c/tests/test_basic_auth.js#L34), passwords are not mandatory.

I couldn't find any specific spec for this, but I guess that if Nock has a test for it someone needed this.

### CODE REVIEW

**Review by kettanaito — APPROVED** [review]
I think you're right. It seems to be possible to construct a URL with `username` alone:

```js
let u = new URL('https://example.com')
u.username = 'user'
u.href
// https://user@example.com/
```

File: src/interceptors/ClientRequest/utils/createRequest.ts [hunk 1]
```diff
@@ -25,7 +25,7 @@ export function createRequest(clientRequest: NodeClientRequest): Request {
    * the request "Authorization" header.
    * @see https://github.com/mswjs/interceptors/issues/438
    */
-  if (clientRequest.url.username && clientRequest.url.password) {
+  if (clientRequest.url.username) {
     const auth = `${clientRequest.url.username}:${clientRequest.url.password}`
```
Comment (kettanaito) [inline src/interceptors/ClientRequest/utils/createRequest.ts:29]: Let's then have a fallback for `clientRequest.url.password` to an empty string. If we don't, when the password is not provided this `auth` string will result in 

```
${username}:undefined
```

And produce a malformed encoded auth string.

**Review by kettanaito — CHANGES_REQUESTED** [review]
This looks great but let's add a fallback to the password value to make sure it's an empty string if it's not set.

**Comment by mikicho** [comment]
image
Seems like password without a user is ok too 😅

File: src/interceptors/ClientRequest/utils/createRequest.ts [hunk 2]
```diff
@@ -25,7 +25,7 @@ export function createRequest(clientRequest: NodeClientRequest): Request {
...
     headers.set('Authorization', `Basic ${btoa(auth)}`)
```
Comment (kettanaito) [inline src/interceptors/ClientRequest/utils/createRequest.ts:30]: Also, it's a good idea to add two more tests covering this:

1. username + no password set. Should result in the encoded `Basic USERNAME:` header.
2. user + password set explicitly to `""`. Should result in the same encoded header.

**Comment by mikicho** [comment]
> add a fallback to the password value to make sure it's an empty string if it's not set.

I thought about it, but isn't it the job of the normalization function?

File: src/interceptors/ClientRequest/utils/createRequest.ts (hunk 2 above)
Comment (mikicho) [reply src/interceptors/ClientRequest/utils/createRequest.ts:30]: I'll add them tomorrow

**Comment by kettanaito** [comment]
@mikicho, not really. The normalization happens between the variable nature of the `ClientRequest` options, but we still treat the given values as-is. In this case, if the `URL` has been given, we will operate with its `url.username` and `url.password` values as they were provided to prevent any confusion. 

Regarding the password being optional too, I suggest then making this check OR-OR and composing the `auth` string taking the optional nature of the two values into account. 

```js
if (username != null || password != null) {
  const resolvedUsername = username || ''
  const resolvedPassword = password || ''
  const auth = `${resolvedUsername}:${resolvedPassword}`
}
```

File: src/interceptors/ClientRequest/utils/createRequest.ts (hunk 2 above)
Comment (kettanaito) [reply src/interceptors/ClientRequest/utils/createRequest.ts:30]: Thank you so much!

**Comment by mikicho** [comment]
@kettanaito PTAL.
Feel free to update the formatting :)

File: src/interceptors/ClientRequest/utils/createRequest.test.ts [hunk 3]
```diff
@@ -83,3 +83,43 @@ it('creates a fetch Request with an empty string body', async () => {
   expect(request.url).toBe('https://api.github.com/')
   expect(request.body).toBe(null)
 })
+
+it('creates a fetch Request with an empty username', async () => {
```
Comment (kettanaito) [inline src/interceptors/ClientRequest/utils/createRequest.test.ts:87]: Looks like the names of these test suites are mixed up: this one is about empty password, and the bottom one is about empty username, isn't it?

Comment (mikicho) [reply src/interceptors/ClientRequest/utils/createRequest.test.ts]: oopsie. fixed 😅

**Review by kettanaito — APPROVED** [review]
Looks great to me. Yet another superb contribution, @mikicho 🚀

File: src/interceptors/ClientRequest/utils/createRequest.test.ts (hunk 3 above)
Comment (kettanaito) [reply src/interceptors/ClientRequest/utils/createRequest.test.ts]: Thank you!

**Comment by kettanaito** [comment]
## Released: v0.25.5 🎉

This has been released in v0.25.5!

- 📄 [**Release notes**](https://github.com/mswjs/interceptors/releases/tag/v0.25.5)
- 📦 [npm package](https://www.npmjs.com/package/@mswjs/interceptors/v/0.25.5)

Make sure to always update to the latest version (`npm i @mswjs/interceptors@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.

**Pull request merged**